src/btree/bt_sync.c

```diff
@@ -109,44 +109,69 @@ __sync_ref_is_obsolete(WT_SESSION_IMPL *session, WT_REF *ref)
 {
     WT_ADDR *addr;
     WT_CELL_UNPACK vpack;
+    WT_PAGE_MODIFY *mod;
+    bool obsolete;
 
     /* Ignore root pages as they can never be deleted. */
-    if (__wt_ref_is_root(ref))
-        return (false);
-
-    /* Guard against marking an in-memory page to be deleted. */
-    if (ref->state != WT_REF_DISK)
+    if (__wt_ref_is_root(ref)) {
+        __wt_verbose(session, WT_VERB_CHECKPOINT_GC, "%p: skipping root page", (void *)ref);
         return (false);
+    }
 
     /* Ignore internal pages, these are taken care of during reconciliation. */
-    if (!__wt_ref_is_leaf(session, ref))
+    if (!__wt_ref_is_leaf(session, ref)) {
+        __wt_verbose(session, WT_VERB_CHECKPOINT_GC, "%p: skipping internal page with parent: %p",
+          (void *)ref, (void *)ref->home);
         return (false);
+    }
 
-    WT_ORDERED_READ(addr, ref->addr);
-    if (!__wt_off_page(ref->home, addr)) {
-        __wt_cell_unpack(session, ref->home, (WT_CELL *)addr, &vpack);
-        return (__wt_txn_visible_all(session, vpack.newest_stop_txn, vpack.newest_stop_ts));
+    obsolete = false;
+    addr = ref->addr;
+    mod = (ref->page != NULL) ? (ref->page->modify) : NULL;
+
+    /* Check for the page obsolete, if the page is modified and reconciled. */
+    if (mod != NULL && mod->rec_result == WT_PM_REC_REPLACE)
+        obsolete = __wt_txn_visible_all(
+          session, mod->mod_replace.newest_stop_txn, mod->mod_replace.newest_stop_ts);
+    else if (mod != NULL && mod->rec_result == WT_PM_REC_MULTIBLOCK)
+        obsolete = __wt_txn_visible_all(
+          session, mod->mod_multi_newest_stop_txn, mod->mod_multi_newest_stop_ts);
+    /* Check for the obsolete from page disk address. */
+    else if (addr != NULL) {
+        if (!__wt_off_page(ref->home, addr)) {
+            __wt_cell_unpack(session, ref->home, (WT_CELL *)addr, &vpack);
+            obsolete = __wt_txn_visible_all(session, vpack.newest_stop_txn, vpack.newest_stop_ts);
+        } else
+            obsolete = __wt_txn_visible_all(session, addr->newest_stop_txn, addr->newest_stop_ts);
```

Comment:
This is a lot of new logic that seems to have been added.  I wonder if a merge didn't work correctly?